### PR TITLE
Added PartialEq to RecordBatch

### LINKED ISSUE #735: Support equality on RecordBatch
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

This would simplify testing. Right now, if I have two record batches that I want to test for equality I need to either look at their individual columns, or convert the batch to a `StructArray`.

**Describe the solution you'd like**

It seems like we should be able to implement `PartialEq` (and possibly `Eq`) for `RecordBatch` to perform equality on the schema and columns.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by matthewmturner**
# Which issue does this PR close?

Closes #735 

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by matthewmturner** [comment]
@bjchambers FYI took a first cut at this.  Let me know your thoughts :)

File: arrow/src/record_batch.rs [hunk 1]
```diff
@@ -419,6 +419,18 @@ impl From<&StructArray> for RecordBatch {
     }
 }
 
+impl PartialEq for RecordBatch {
+    fn eq(&self, other: &RecordBatch) -> bool {
+        let batches_equal = true;
```
Comment (alamb) [inline arrow/src/record_batch.rs:424]: You might be able to express this in a more idomatic Rust with something like

```rust
self
  .columns()
  .iter()
  .zip(other.columns.iter())
  .map(|(my_arr, other_arr)| my_arr == other.arr)
  .all()
```

Or something similar.

**Review by alamb — APPROVED** [review]
Thank you @matthewmturner  - looks good to me!

**Comment by alamb** [comment]
I am sorry for the delay in review -- I have been away for a while and am now catching up

**Comment by matthewmturner** [comment]
> I am sorry for the delay in review -- I have been away for a while and am now catching up

No apology necessary :)

File: arrow/src/record_batch.rs (hunk 1 above)
Comment (matthewmturner) [reply arrow/src/record_batch.rs:424]: Ok! I will update.  I appreciate this type of feedback as im relatively new to Rust, and still learning to write idiomatic Rust.  Thanks!

**Comment by jorgecarleitao** [comment]
Can't we add `PartialEq` to `#derive()`? It seems to me that this PR ignores differences in the schema (e.g. column names)

**Comment by alamb** [comment]
Checking for different column names is an excellent point @jorgecarleitao

**Comment by matthewmturner** [comment]
Thx both - I will add check for column names.

@jorgecarleitao regarding adding the `PartialEq` to `#derive()` - i had looked into this but saw the following in the docs (https://doc.rust-lang.org/std/cmp/trait.PartialEq.html):

Specifically: 
"This trait can be used with #[derive]. When derived on structs, two instances are equal if all fields are equal, and not equal if any fields are not equal. When derived on enums, each variant is equal to itself and not equal to the other variants."

I didnt think this would translate to the `RecordBatch` struct well - but tbh i didnt try.  I can give it a shot though and see if it works.

let me know if im misunderstanding anything.

**Comment by alamb** [comment]
@matthewmturner  -- I think trying to add `#[derive(PartialEq)]` on `RecordBatch` is probably a good place to start -- it would compare two `RecordBatch`es field by field which seems to be exactly what we would want

**Review by nevi-me — CHANGES_REQUESTED** [review]
Hi @matthewmturner the equality check will generate false positives as is, because if `self` has 2 columns but `other` has n where `n > 2` and the first 2 columns are the same; we'd incorrectly say that the batches are equal.

I'd propose a solution that checks the dimension of the batches first, inaddition to what @alamb has suggested.

Something like

```rust
impl ParialEq for RecordBatch {
  fn eq(&self, other: &RecordBatch) -> bool {
    // check that the width and length of the two batches are the same, then check each column
    self.num_columns() == other.num_columns() && self.num_rows() == other.num_rows()
    && self.columns()
      .iter()
      .zip(other.columns.iter())
      .all(|(a, b)| a == b)
  }
}
```

___

Then on @jorgecarleitao's comment re schema equality, I think we could have two methods, one that checks that batch + schema are equal, and another that only checks batches.

I say this because we have a relaxation in creating batches, where we use `RecordBatchOptions` in `RecordBatch::try_new_with_options()`, which could lead to a default method that checks schema + batch failing if batches are created with the `try_new_with_options` call.

We could look for guidance on the Arrow format docs (or look at other implementations), as I'm unsure of what the correct default implementation should be.

At the least, the implementation on this PR can just check the dimensions + individual column equality.

**Comment by matthewmturner** [comment]
@nevi-me thank you for the feedback.

I've added tests to cover the additional RecordBatch differences (col names, col numbers, row numbers).  Also, i tried just adding `PartialEq` to `#derive()` as @jorgecarleitao suggested and it seems to have worked as all tests pass.  So i dont think we need a custom implementation for `PartialEq` now.

Let me know if this covers a good scope for this PR.  If so, I can create a new issue to look into the point by @nevi-me on `RecordBatchOptions`

**Review by nevi-me — APPROVED** [review]

**Comment by nevi-me** [comment]
CI failures seem to be related to a packed_simd_2 release from 15 hours ago

**Comment by matthewmturner** [comment]
@nevi-me does that mean i just need to rebase?

**Comment by alamb** [comment]
> @nevi-me does that mean i just need to rebase?


It probably means we need to sort out what broke first and fix it.

**Comment by alamb** [comment]
Filed https://github.com/apache/arrow-rs/pull/765 to track packed simd failure

**Comment by matthewmturner** [comment]
@alamb @nevi-me i rebased and CI clears now.

**Review by alamb — COMMENTED** [review]
The changes to `RecordBatch` looks good to me @matthewmturner  -- for some reason this PR also appears to contain changes from several other commits. I wonder if something went wrong with the rebase...

**Comment by matthewmturner** [comment]
@alamb i think i fixed...i was expecting to have to rebase again after resetting, but all tests are passing now.  let me know if anything seems off to you.

**Review by nevi-me — APPROVED** [review]

**Comment by alamb** [comment]
Thanks for sticking with this @matthewmturner  and the great comments @nevi-me  and @jorgecarleitao 👍

**Comment by matthewmturner** [comment]
@alamb @jorgecarleitao @nevi-me thanks all for guidance!

**Pull request merged**